src/ray/gcs/gcs_server/gcs_resource_scheduler.cc

```diff
@@ -121,12 +142,70 @@ absl::flat_hash_set<NodeID> GcsResourceScheduler::FilterCandidateNodes(
   return result;
 }
 
-const std::vector<ResourceSet> &GcsResourceScheduler::SortRequiredResources(
+std::tuple<bool, bool> cmp_resource(
+    const std::string &resource,
+    const absl::flat_hash_map<std::string, FixedPoint> &a_map,
+    const absl::flat_hash_map<std::string, FixedPoint> &b_map) {
+  auto a_resource = a_map.find(resource);
+  auto b_resource = b_map.find(resource);
+
+  bool a_less_than_b = false;
+  bool finished = false;
+
+  if (a_resource != a_map.end() && b_resource != b_map.end()) {
+    if (a_resource->second != b_resource->second) {
+      a_less_than_b = a_resource->second < b_resource->second;
+      finished = true;
+    }
+  } else if (a_resource != a_map.end() && a_resource->second != 0) {
+    a_less_than_b = false;
+    finished = true;
+  } else if (b_resource != b_map.end() && b_resource->second != 0) {
+    a_less_than_b = true;
+    finished = true;
+  }
+  return std::make_tuple(finished, a_less_than_b);
+}
+
+std::vector<size_t> GcsResourceScheduler::SortRequiredResources(
     const std::vector<ResourceSet> &required_resources) {
-  // TODO(ffbin): A bundle may require special resources, such as GPU. We need to
-  // schedule bundles with special resource requirements first, which will be implemented
-  // in the next pr.
-  return required_resources;
+  std::vector<size_t> sorted_index(required_resources.size());
+  std::iota(sorted_index.begin(), sorted_index.end(), 0);
+  std::sort(sorted_index.begin(), sorted_index.end(), [required_resources](int b, int a) {
+    auto a_map = required_resources[a].GetResourceAmountMap();
+    auto b_map = required_resources[b].GetResourceAmountMap();
+    bool finished, cmp_res;
+    // TODO (jon-chuang): the exact resource priority defined here needs to be revisted.
+    std::tie(finished, cmp_res) = cmp_resource("GPU", a_map, b_map);
```

Comment:
This part is a little hard to understand. Can you tell me high level description?

Reply:
This just how to implement a comparator for sorting in c++. We return true if a < b. Here we are comparing two maps to see if a_map's resource priority < b_maps, based on the resource "GPU". If we can resolve it, we return finished = true. Else, we rely on the resource of the next priority level to try to resolve, all the way until "CPU" after which we return false.

Reply:
I've added comments